internal/rule/squashjoins.go

```diff
@@ -0,0 +1,948 @@
+	case t1 == gitquery.ReferencesTableName && t2 == gitquery.BlobsTableName:
+		return isCommitHasBlob(
+			isCol(gitquery.ReferencesTableName, "hash"),
+			isCol(gitquery.BlobsTableName, "hash"),
+		)(f)
+	case t1 == gitquery.CommitsTableName && t2 == gitquery.TreeEntriesTableName:
+		return isEq(
+			isCol(gitquery.CommitsTableName, "tree_hash"),
+			isCol(gitquery.TreeEntriesTableName, "tree_hash"),
+		)(f) || isCommitHasTree(
+			isCol(gitquery.CommitsTableName, "hash"),
+			isCol(gitquery.TreeEntriesTableName, "tree_hash"),
+		)(f) || isCommitHasBlob(
+			isCol(gitquery.CommitsTableName, "hash"),
+			isCol(gitquery.TreeEntriesTableName, "entry_hash"),
+		)(f)
+	case t1 == gitquery.CommitsTableName && t2 == gitquery.BlobsTableName:
+		return isCommitHasBlob(
+			isCol(gitquery.CommitsTableName, "hash"),
+			isCol(gitquery.BlobsTableName, "hash"),
+		)(f)
+	case t1 == gitquery.TreeEntriesTableName && t2 == gitquery.BlobsTableName:
+		return isEq(
+			isCol(gitquery.TreeEntriesTableName, "entry_hash"),
+			isCol(gitquery.BlobsTableName, "hash"),
+		)(f)
+	}
+	return false
+}
+
+type validator func(sql.Expression) bool
+
+func isEq(left, right validator) validator {
+	return func(e sql.Expression) bool {
+		eq, ok := e.(*expression.Equals)
+		if !ok {
+			return false
+		}
+
+		return (left(eq.Left()) && right(eq.Right())) ||
+			(right(eq.Left()) && left(eq.Right()))
+	}
+}
+
+func isCol(table, name string) validator {
+	return func(e sql.Expression) bool {
+		gf, ok := e.(*expression.GetField)
+		if !ok {
+			return false
+		}
+
+		return gf.Table() == table && gf.Name() == name
+	}
+}
+
+func isCommitHasTree(left, right validator) validator {
+	return func(e sql.Expression) bool {
+		f, ok := e.(*function.CommitHasTree)
+		if !ok {
+			return false
+		}
+
+		return left(f.Left) && right(f.Right)
+	}
+}
+
+func isCommitHasBlob(left, right validator) validator {
+	return func(e sql.Expression) bool {
+		f, ok := e.(*function.CommitHasBlob)
+		if !ok {
+			return false
+		}
+
+		return left(f.Left) && right(f.Right)
+	}
+}
+
+func isHistoryIdx(left, right validator) validator {
+	return func(e sql.Expression) bool {
+		f, ok := e.(*function.HistoryIdx)
+		if !ok {
+			return false
+		}
+
+		return left(f.Left) && right(f.Right)
+	}
+}
+
+func isGte(left, right validator) validator {
+	return func(e sql.Expression) bool {
+		switch e := e.(type) {
+		case *expression.GreaterThanOrEqual:
+			return left(e.Left()) && right(e.Right())
+		case *expression.LessThanOrEqual:
+			return left(e.Right()) && right(e.Left())
+		default:
+			return false
+		}
+	}
+}
+
+func isNum(n int64) validator {
+	return func(e sql.Expression) bool {
+		lit, ok := e.(*expression.Literal)
+		if !ok {
+			return false
+		}
+
+		result, err := lit.Eval(nil, nil)
+		if err != nil {
+			return false
+		}
+
+		num, ok := result.(int64)
+		if !ok {
+			return false
+		}
+
+		return num == n
+	}
+}
+
+var errColumnNotFound = errors.NewKind("column %s.%s not found in schema")
```

Comment:
we should use ErrColumnNotFound from `go-mysql-server`: https://sourcegraph.com/github.com/src-d/go-mysql-server/-/blob/sql/analyzer/rules.go#L32